xmtp_mls/src/groups/validated_commit_v2.rs

```diff
@@ -0,0 +1,458 @@
+};
+use prost::Message;
+use thiserror::Error;
+#[cfg(doc)]
+use xmtp_id::associations::AssociationState;
+use xmtp_proto::{
+    api_client::{XmtpIdentityClient, XmtpMlsClient},
+    xmtp::identity::MlsCredential,
+};
+
+use crate::{
+    configuration::GROUP_MEMBERSHIP_EXTENSION_ID,
+    identity_updates::{InstallationDiff, InstallationDiffError},
+    storage::db_connection::DbConnection,
+    Client,
+};
+
+use super::{
+    group_membership::{GroupMembership, MembershipDiff},
+    group_metadata::{extract_group_metadata, GroupMetadataError},
+};
+
+#[derive(Debug, Error)]
+pub enum CommitValidationError {
+    #[error("Actor could not be found")]
+    ActorCouldNotBeFound,
+    // Subject of the proposal has an invalid credential
+    #[error("Inbox validation failed for {0}")]
+    InboxValidationFailed(String),
+    // Not used yet, but seems obvious enough to include now
+    #[error("Insufficient permissions")]
+    InsufficientPermissions,
+    // TODO: We will need to relax this once we support external joins
+    #[error("Actor not a member of the group")]
+    ActorNotMember,
+    #[error("Subject not a member of the group")]
+    SubjectDoesNotExist,
+    // Current behaviour is to error out if a Commit includes proposals from multiple actors
+    // TODO: We should relax this once we support self remove
+    #[error("Multiple actors in commit")]
+    MultipleActors,
+    #[error("Missing group membership")]
+    MissingGroupMembership,
+    #[error("Unexpected installations added: {0:?}")]
+    UnexpectedInstallationAdded(Vec<Vec<u8>>),
+    #[error("Sequence ID can only increase")]
+    SequenceIdDecreased,
+    #[error("Unexpected installations removed: {0:?}")]
+    UnexpectedInstallationsRemoved(Vec<Vec<u8>>),
+    #[error(transparent)]
+    GroupMetadata(#[from] GroupMetadataError),
+    #[error(transparent)]
+    MlsCredential(#[from] BasicCredentialError),
+    #[error(transparent)]
+    ProtoDecode(#[from] prost::DecodeError),
+    #[error(transparent)]
+    InstallationDiff(#[from] InstallationDiffError),
+}
+
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub(crate) struct CommitParticipant {
+    pub inbox_id: String,
+    pub installation_id: Vec<u8>,
+    pub is_creator: bool,
+    // TODO: Add is_admin
+}
+
+impl TryFrom<&LeafNode> for CommitParticipant {
+    type Error = CommitValidationError;
+
+    fn try_from(leaf_node: &LeafNode) -> Result<Self, Self::Error> {
+        Ok(Self {
+            inbox_id: inbox_id_from_credential(leaf_node.credential())?,
+            installation_id: leaf_node.signature_key().as_slice().to_vec(),
+            is_creator: false,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct Inbox {
+    pub inbox_id: String,
+    pub is_creator: bool,
+    // TODO: add is_admin support
+    // pub is_admin: bool,
+}
+
+/**
+ * A [`ValidatedCommit`] is a summary of changes coming from a MLS commit, after all of our validation rules have been applied
+ *
+ * Commit Validation Rules:
```

Comment:
I tried to boil the validations in XIP-46 down to these 5 rules. They don't include the rules around permissions (who is actually allowed to add/remove members or mutate metadata), which will be handled separately.